src/indexer.c

```diff
@@ -0,0 +1,184 @@
+#include <string.h>
+#include <rmutil/alloc.h>
+#include <rmutil/vector.h>
+
+#include "consts.h"
+#include "indexer.h"
+
+int parseLabel(RedisModuleCtx *ctx, RedisModuleString *label, Label *retLabel, const char *separator) {
+    char *token;
+    char *iter_ptr;
+    size_t _s;
+    const char *labelRaw = RedisModule_StringPtrLen(label, &_s);
+    char *labelstr = RedisModule_PoolAlloc(ctx, _s + 1);
+    labelstr[_s] = '\0';
+    strncpy(labelstr, labelRaw, _s);
+    token = strtok_r(labelstr, separator, &iter_ptr);
+    for (int i=0; i<2; i++)
+    {
+        if (token == NULL && i==0) {
+            return TSDB_ERROR;
+        }
+        if (i == 0) {
+            retLabel->key = RedisModule_CreateString(NULL, token, strlen(token));
+        } else {
+            if (token != NULL) {
+                retLabel->value = RedisModule_CreateString(NULL, token, strlen(token));
+            } else {
+                retLabel->value = NULL;
+            }
+        }
+        token = strtok_r (NULL, separator, &iter_ptr);
+    }
+    return TSDB_OK;
+}
+
+void IndexOperation(RedisModuleCtx *ctx, const char *op, RedisModuleString *ts_key, Label *labels, size_t labels_count) {
+    RedisModuleCallReply *reply;
+
+    for (int i=0; i<labels_count; i++) {
+        size_t _s;
+        RedisModuleString *indexed_key_value = RedisModule_CreateStringPrintf(ctx, "__index_%s=%s",
+                RedisModule_StringPtrLen(labels[i].key, &_s),
+                RedisModule_StringPtrLen(labels[i].value, &_s));
+        RedisModuleString *indexed_key = RedisModule_CreateStringPrintf(ctx, "__index_%s",
+                RedisModule_StringPtrLen(labels[i].key, &_s),
+                RedisModule_StringPtrLen(labels[i].value, &_s));
+        reply = RedisModule_Call(ctx, op, "ss", indexed_key_value, ts_key);
+        if (reply)
+            RedisModule_FreeCallReply(reply);
+        reply = RedisModule_Call(ctx, op, "ss", indexed_key, ts_key);
+        if (reply)
+            RedisModule_FreeCallReply(reply);
+    }
+}
+
+void IndexMetric(RedisModuleCtx *ctx, RedisModuleString *ts_key, Label *labels, size_t labels_count) {
+    IndexOperation(ctx, "SADD", ts_key, labels, labels_count);
+}
+
+void RemoveIndexedMetric(RedisModuleCtx *ctx, RedisModuleString *ts_key, Label *labels, size_t labels_count) {
+    IndexOperation(ctx, "SREM", ts_key, labels, labels_count);
+}
+
+int _intersect(RedisModuleDict *left, RedisModuleDict *right, char **lastKey, size_t *lastKeySize) {
+    RedisModuleDictIter *iter;
+    if (*lastKey == NULL) {
+        iter = RedisModule_DictIteratorStartC(left, "^", NULL, 0);
+    } else {
+        iter = RedisModule_DictIteratorStartC(left, "^", lastKey, *lastKeySize);
```

Comment:
I'm not sure what this line is trying to do, but RedisModule_DictIteratorStartC with "^" ignores the given key